### PR TITLE
Refactor bonded interaction parameter handling

### LINKED ISSUE #3019: Bond_parameters should be a variant
This is currently a tagged union, which requires manual construction and destruction when changing the active type (the union contains non-trivial types). This requires boiler plate code for switching the active type and is error prone. This can be made automatic by turning the union
into a `boost::variant`.

### PR DESCRIPTION
**Pull request by biermanncarl**
Fixes #3019, partial fix for #3680, closes #4066

Description of changes:
- Bonded parameters are now stored inside a ```boost::variant```
- Most bond types are now in self-contained header files (dependency reversals compared to former state)
- Most bond kernels have become members of the respective structs
- Each bond type is represented by its own C++ type (the three tabulated types are now separate, too), which is important because C++ type is now equivalent to bond type
- Removed the former bond type identification scheme from the C++ core
- Separate bonded IA cutoff from non-bonded IA cutoff
- Skip bond loop independently from non-bonded loop

### CODE REVIEW

File: src/config/config.hpp [hunk 1]
```diff
@@ -90,6 +90,11 @@
 #define TINY_OIF_ELASTICITY_COEFFICIENT 1e-10
 #endif
 
+/** Maximal number of soft particles per IBM volcons bond (?) */
+#ifndef IBM_MAX_NUM
+#define IBM_MAX_NUM 1000
```
Comment (KaiSzuttor) [inline src/config/config.hpp]: why?

File: src/core/bonded_interactions/angle_cosine.hpp [hunk 2]
```diff
@@ -67,18 +98,16 @@ angle_cosine_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
...
+    Utils::Vector3d const &r_right) const {
   auto const vectors = calc_vectors_and_cosine(r_mid, r_left, r_right, true);
   auto const cos_phi = std::get<4>(vectors);
   auto const sin_phi = sqrt(1 - Utils::sqr(cos_phi));
-  auto const cos_phi0 = iaparams.p.angle_cosine.cos_phi0;
-  auto const sin_phi0 = iaparams.p.angle_cosine.sin_phi0;
-  auto const k = iaparams.p.angle_cosine.bend;
+  auto const cos_phi0 = this->cos_phi0;
+  auto const sin_phi0 = this->sin_phi0;
+  auto const k = this->bend;
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_cosine.hpp]: these variables are not needed

File: src/core/bonded_interactions/angle_cossquare.hpp [hunk 3]
```diff
@@ -27,32 +27,59 @@
...
+Angle_cossquare_bond_parameters::angle_force(
+    Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
+    Utils::Vector3d const &r_right) const {
 
-  auto forceFactor = [&iaparams](double const cos_phi) {
-    auto const cos_phi0 = iaparams.p.angle_cossquare.cos_phi0;
-    auto const k = iaparams.p.angle_cossquare.bend;
+  auto forceFactor = [this](double const cos_phi) {
+    auto const cos_phi0 = this->cos_phi0;
+    auto const k = this->bend;
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_cossquare.hpp]: these variables are not needed

File: src/core/bonded_interactions/angle_cossquare.hpp [hunk 4]
```diff
@@ -63,16 +90,14 @@ angle_cossquare_force(Utils::Vector3d const &r_mid,
...
-                                     Bonded_ia_parameters const &iaparams) {
+inline double Angle_cossquare_bond_parameters::angle_energy(
+    Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
+    Utils::Vector3d const &r_right) const {
   auto const vectors = calc_vectors_and_cosine(r_mid, r_left, r_right, true);
   auto const cos_phi = std::get<4>(vectors);
-  auto const cos_phi0 = iaparams.p.angle_cossquare.cos_phi0;
-  auto const k = iaparams.p.angle_cossquare.bend;
+  auto const cos_phi0 = this->cos_phi0;
+  auto const k = this->bend;
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_cossquare.hpp]: here and elsewhere, please remove those variables, also you do not need to use "this" explicitely, I think we use  the convention of using a "m_" prefix for member variables

File: src/core/bonded_interactions/bonded_coulomb.hpp [hunk 5]
```diff
@@ -29,51 +29,67 @@
...
 #include <cmath>
 
-/** Set the parameters for the bonded Coulomb potential
- *
- *  @retval ES_OK on success
- *  @retval ES_ERROR on error
- */
-int bonded_coulomb_set_params(int bond_type, double prefactor);
+/** Parameters for %Coulomb bond Potential */
+struct Bonded_coulomb_bond_parameters {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb.hpp]:
```suggestion
struct BondedCoulombParameters {
```

File: src/core/bonded_interactions/bonded_interaction_data.cpp [hunk 6]
```diff
@@ -28,68 +28,29 @@
...
+};
 
 double maximal_cutoff_bonded() {
-  auto const max_cut_bonded =
-      boost::accumulate(bonded_ia_params, -1.,
-                        [](auto max_cut, Bonded_ia_parameters const &bond) {
-                          return std::max(max_cut, cutoff(bond.type, bond.p));
-                        });
+  auto const max_cut_bonded = boost::accumulate(
+      bonded_ia_params, -1.,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_interaction_data.cpp]: why do you initialize with -1 and not 0?

File: src/core/bonded_interactions/bonded_interaction_data.cpp [hunk 7]
```diff
@@ -99,31 +60,28 @@ double maximal_cutoff_bonded() {
   return (any_dihedrals) ? 2 * max_cut_bonded : max_cut_bonded;
 }
 
+int bonded_ia_params_num_partners(int bond_id) {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_interaction_data.cpp]:
```suggestion
int no_of_partners(int bond_id) {
```

File: src/config/config.hpp (hunk 1 above)
Comment (biermanncarl) [reply src/config/config.hpp]: It's basically a relocation of a constant from [here](https://github.com/espressomd/espresso/pull/4161/files#diff-17cc54295fcd4d75a78d8790cb5c73e2330c7f30d7a504f86ca89c16513bf527L30)

File: src/core/bonded_interactions/angle_cossquare.hpp (hunk 4 above)
Comment (biermanncarl) [reply src/core/bonded_interactions/angle_cossquare.hpp]: Well, it's a tradeoff. Either we change the name of ALL bonded parameter members, or we keep ```this->```. It's actually only needed in the constructor, and nowhere else if I am not mistaken. I'm unsure what is the better solution.

Comment (KaiSzuttor) [reply src/core/bonded_interactions/angle_cossquare.hpp]: ```this``` is only ever needed for disambiguation. And if you are touching those variables we should change the naming to be consistent with our conventions

Comment (jngrad) [reply src/core/bonded_interactions/angle_cossquare.hpp]: There is no consensus in core team yet for adding a prefix `m_` to public members. The `this->` dereference however can be removed in all bonded force/energy kernels.

File: src/core/bonded_interactions/bonded_interaction_data.cpp (hunk 6 above)
Comment (jngrad) [reply src/core/bonded_interactions/bonded_interaction_data.cpp]: I think `-1.` is the magic value that will be replaced by the more expressive macro `BONDED_INACTIVE_CUTOFF` in #4066.

File: src/core/bonded_interactions/bonded_interaction_data.hpp [hunk 8]
```diff
@@ -23,349 +23,114 @@
...
+/** Field containing the parameters of the bonded ia types */
+extern std::vector<Bonded_ia_parameters> bonded_ia_params;
 
-/** Parameters for n-body tabulated potential (n=2,3,4). */
-struct Tabulated_bond_parameters {
-  TabulatedBondedInteraction type;
-  TabulatedPotential *pot;
+/** These following functions are necessary to hide boost::variant from Cython.
+ *  It can't be used there because Cython doesn't support variadic templates.
+ */
```
Comment (jngrad) [inline src/core/bonded_interactions/bonded_interaction_data.hpp]: It's unclear to which function this doxygen block applies. You can use `@name` and delimiters `/**@{*/` and `/**@}*/` to convey that information ([example](https://github.com/espressomd/espresso/blob/a20d99b94f9ee8c798afdc56f83d51858206beb2/src/core/thermostat.hpp#L58-L67)).

File: src/core/bonded_interactions/bonded_interactions.dox [hunk 9]
```diff
@@ -24,131 +24,164 @@
...
- *  * core:
- *   - define a data structure for the interaction parameters (prefactors, etc.)
- *   - write a setter function, which takes the parameters of the interactions
- *     and stores them in the interaction data structure
- *   - write functions for calculating force and energy
- *   - add calls to the force and energy calculation functions to the force
- *     calculation in the integration loop as well as to energy and
- *     pressure tensor analysis
+ *  * C++ core:
+ *    - Create a new header filefor your new bond type, preferably inside the
```
Comment (jngrad) [inline src/core/bonded_interactions/bonded_interactions.dox]:
```suggestion
 *    - Create a new header file for your new bond type, preferably inside the
```

File: src/core/bonded_interactions/bonded_interactions.dox [hunk 10]
```diff
@@ -24,131 +24,164 @@
...
- *  * \c bonded_ia_params[bond_type].num is the number of particles involved
- *    in the bond, minus 1, i.e. 1 for a pairwise bonded potential such as the
- *    FENE bond.
- *  * The parameters for the individual bonded interaction go to the member of
- *    @ref Bond_parameters for your interaction defined in the previous step.
- *    For the FENE bond, this would be:
+ *    Replace all the <tt>ar &</tt> commands with the new bond's parameters.
+ *    Every data member of your struct needs to be transmitted. This template
+ *    function can be made private, but then the following line needs to be
+ *    added before for the serialization routine to be able to access it:
```
Comment (jngrad) [inline src/core/bonded_interactions/bonded_interactions.dox]: serialization methods are typically private, therefore I would rephrase it like so:
```suggestion
 *    Every data member of your struct needs to be transmitted. This template
 *    function is private, hence the following line needs to be added before:
```

File: src/core/bonded_interactions/bonded_interactions.dox [hunk 11]
```diff
@@ -24,131 +24,164 @@
...
- *  * The functions calculating force and energy can make use of a
- *    pre-calculated distance vector (\p dx) pointing from particle 2 to
- *    particle 1, that takes periodic boundary conditions into account.
- *  * The force and energy functions cannot alter the particle states.
- *
- *  @subsection bondedIA_new_integration Including the bonded interaction in \
- *                                       the force calculation and the energy \
- *                                       and pressure analysis
+ *    Refer to @ref fene.hpp as an example.
+
```
Comment (jngrad) [inline src/core/bonded_interactions/bonded_interactions.dox]:
```suggestion
```

File: src/core/bonded_interactions/bonded_interactions.dox [hunk 12]
```diff
@@ -503,3 +533,58 @@
  *  @f$ \theta_{ijk} = \pi @f$.
  *
  */
+
+/*
```
Comment (jngrad) [inline src/core/bonded_interactions/bonded_interactions.dox]: This break transformed the next block into a regular comment instead of a doxygen comment.
```suggestion
 *
```

File: src/python/espressomd/interactions.pxd [hunk 13]
```diff
@@ -298,45 +298,97 @@ IF SOFT_SPHERE:
...
-        int type
-        TabulatedPotential * pot
+    cdef extern from "bonded_interactions/bonded_interaction_data.hpp":
+        # Parameters for n-body tabulated potential (n=2,3,4).
+        cdef cppclass Tabulated_bond_parameters:
+            Tabulated_bond_parameters()
+            Tabulated_bond_parameters(double min, double max,
+                                      const vector[double] & energy,
+                                      const vector[double] & force) except +
+            int type
```
Comment (jngrad) [inline src/python/espressomd/interactions.pxd]:
```suggestion
```

File: src/python/espressomd/interactions.pxd [hunk 14]
```diff
@@ -298,45 +298,97 @@ IF SOFT_SPHERE:
...
+            TabulatedPotential * pot
```
Comment (jngrad) [inline src/python/espressomd/interactions.pxd]: Should this raw pointer become a `shared_ptr[TabulatedPotential] pot` with the appropriate `from libcpp.memory cimport shared_ptr` at the top of the Cython file?

File: src/config/config.hpp (hunk 1 above)
Comment (jngrad) [reply src/config/config.hpp]: This constant had to be moved out to help with dependency inversion, IIRC.

File: src/core/bonded_interactions/bonded_interaction_data.cpp (hunk 7 above)
Comment (jngrad) [reply src/core/bonded_interactions/bonded_interaction_data.cpp]: The `bonded_ia_params_` prefix is used in this function and the ones above and below because it's acting on the variant called `bonded_ia_params`. This will become clear once we add the correct doxygen group delimiters. We could even move these functions to `src/script_interface`.

File: src/core/bonded_interactions/bonded_coulomb.hpp (hunk 5 above)
Comment (jngrad) [reply src/core/bonded_interactions/bonded_coulomb.hpp]: All bonded parameter structs are now CamelCase.

File: src/core/bonded_interactions/bonded_interaction_data.cpp (hunk 7 above)
Comment (jngrad) [reply src/core/bonded_interactions/bonded_interaction_data.cpp]: Function moved to `src/script_interface/interactions/bonded.hpp` and modified to call a new function with signature `number_of_partners(Bonded_ia_parameters const &iaparams)` in `src/core/bonded_interactions/bonded_interaction_data.cpp`.

**Review by jngrad — DISMISSED** [review]
good to go from my side

File: src/core/bonded_interactions/angle_cosine.hpp [hunk 15]
```diff
@@ -27,37 +27,65 @@
...
+
+  double cutoff() const { return -1.; }
+
+  static constexpr int num = 2;
+
+  AngleCosineBondParameters() = default;
+  AngleCosineBondParameters(double bend, double phi0);
+
+  std::tuple<Utils::Vector3d, Utils::Vector3d, Utils::Vector3d>
+  angle_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_cosine.hpp]: i'm not sure about ```angle_force```... wouldn't ```forces``` more precisely reflect that there are multiple forces returned and calculated?

File: src/core/bonded_interactions/angle_cosine.hpp [hunk 16]
```diff
@@ -67,21 +95,17 @@ angle_cosine_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
...
  *  @param[in]  r_left    Position of first/left particle.
  *  @param[in]  r_right   Position of third/right particle.
- *  @param[in]  iaparams  Bonded parameters for the angle interaction.
  */
-inline double angle_cosine_energy(Utils::Vector3d const &r_mid,
-                                  Utils::Vector3d const &r_left,
-                                  Utils::Vector3d const &r_right,
-                                  Bonded_ia_parameters const &iaparams) {
+inline double
+AngleCosineBondParameters::angle_energy(Utils::Vector3d const &r_mid,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_cosine.hpp]:
```suggestion
AngleCosineBondParameters::energy(Utils::Vector3d const &r_mid,
```

File: src/core/bonded_interactions/angle_harmonic.hpp [hunk 17]
```diff
@@ -27,36 +27,58 @@
...
 #include <utils/Vector.hpp>
 #include <utils/math/sqr.hpp>
 
 #include <cmath>
 #include <tuple>
 
-/** Set parameters for the angle potential. */
-int angle_harmonic_set_params(int bond_type, double bend, double phi0);
+/** Parameters for three-body angular potential (harmonic). */
+struct AngleHarmonicBondParameters {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/angle_harmonic.hpp]: why are these types have a ```Parameters``` suffix if they actually also implement the energy and force kernels?

File: src/core/bonded_interactions/bonded_coulomb_sr.hpp [hunk 18]
```diff
@@ -42,40 +39,58 @@
...
+  double q1q2;
+
+  double cutoff() const { return -1.; }
+
+  static constexpr int num = 1;
+
+  BondedCoulombSRParameters() = default;
+  BondedCoulombSRParameters(double q1q2);
+
+  boost::optional<Utils::Vector3d> pair_force(Utils::Vector3d const &dx) const;
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb_sr.hpp]:
```suggestion
  boost::optional<Utils::Vector3d> force(Utils::Vector3d const &dx) const;
```

File: src/core/bonded_interactions/bonded_coulomb_sr.hpp [hunk 19]
```diff
@@ -42,40 +39,58 @@
...
+  boost::optional<double> pair_energy(Particle const &p1, Particle const &p2,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb_sr.hpp]:
```suggestion
  boost::optional<double> energy(Particle const &p1, Particle const &p2,
```

File: src/core/bonded_interactions/bonded_interaction_data.cpp (hunk 6 above)
Comment (KaiSzuttor) [reply src/core/bonded_interactions/bonded_interaction_data.cpp]: Still don't understand this... The minimum value for the maximum cut off is zero and not some magic number.

File: src/core/bonded_interactions/bonded_interaction_data.hpp [hunk 20]
```diff
@@ -23,346 +23,80 @@
...
-  BondedInteraction type;
-  /** Number of bonds (N_part - 1) for that interaction type. */
-  int num;
-  /** Interaction parameters. */
-  Bond_parameters p;
-};
+/** Variant in which to store the parameters of an individual bonded
+ *  interaction
+ */
+using Bonded_ia_parameters = boost::variant<
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_interaction_data.hpp]:
```suggestion
using Bonded_IA_Parameters = boost::variant<
```

File: src/core/bonded_interactions/bonded_tab.cpp [hunk 21]
```diff
@@ -21,66 +21,54 @@
...
 #include <vector>
 
-int tabulated_bonded_set_params(int bond_type,
-                                TabulatedBondedInteraction tab_type, double min,
-                                double max, std::vector<double> const &energy,
-                                std::vector<double> const &force) {
-  if (bond_type < 0)
-    return ES_ERROR;
-
+Tabulated_bond_parameters::Tabulated_bond_parameters(
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_tab.cpp]:
```suggestion
TabulatedBondParameters::TabulatedBondParameters(
```

File: src/core/bonded_interactions/bonded_tab.hpp [hunk 22]
```diff
@@ -32,53 +32,112 @@
...
- *
- *  @retval ES_OK on success
- *  @retval ES_ERROR on error
- */
-int tabulated_bonded_set_params(int bond_type,
-                                TabulatedBondedInteraction tab_type, double min,
-                                double max, std::vector<double> const &energy,
-                                std::vector<double> const &force);
+/** Base class for n-body tabulated potential (n=2,3,4). */
+struct Tabulated_bond_parameters {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_tab.hpp]:
```suggestion
struct TabulatedBonParameters {
```

File: src/core/bonded_interactions/bonded_tab.hpp [hunk 23]
```diff
@@ -32,53 +32,112 @@
...
+
+  static constexpr int num = 3;
+
+  TabulatedDihedralBondParameters() = default;
+  TabulatedDihedralBondParameters(double min, double max,
+                                  std::vector<double> const &energy,
+                                  std::vector<double> const &force);
+  boost::optional<std::tuple<Utils::Vector3d, Utils::Vector3d, Utils::Vector3d,
+                             Utils::Vector3d>>
+  dihedral_force(Utils::Vector3d const &r1, Utils::Vector3d const &r2,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_tab.hpp]:
```suggestion
  forces(Utils::Vector3d const &r1, Utils::Vector3d const &r2,
```

File: src/core/bonded_interactions/bonded_tab.hpp [hunk 24]
```diff
@@ -32,53 +32,112 @@
...
  *  extrapolation based on the first two tabulated force values.
  *
- *  @param[in]  iaparams  Bonded parameters for the pair interaction.
  *  @param[in]  dx        %Distance between the particles.
  */
 inline boost::optional<Utils::Vector3d>
-tab_bond_force(Bonded_ia_parameters const &iaparams,
-               Utils::Vector3d const &dx) {
-  auto const *tab_pot = iaparams.p.tab.pot;
+TabulatedDistanceBondParameters::pair_force(Utils::Vector3d const &dx) const {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_tab.hpp]:
```suggestion
TabulatedDistanceBond::force(Utils::Vector3d const &dx) const {
```

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
Please check that all types follow our naming convention (SnakeCase) and adjust the naming of the bond classes (remove the ```Parameters``` suffix)

**Review by KaiSzuttor — APPROVED** [review]
